kubemark：--use-real-proxier=true will cause panic runtime error

<!-- This form is for bug reports and feature requests ONLY!

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).

If the matter is security related, please disclose it privately via https://kubernetes.io/security/.
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>

/kind bug

> /kind feature


**What happened**:

when i run hollow-node deployment，hollow-proxy will cause panic.

hollow-node deployment is as follows:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "10"
  creationTimestamp: 2018-10-12T09:42:16Z
  generation: 10
  labels:
    name: hollow-node
  name: hollow-node
  namespace: kubemark1
  resourceVersion: "14267172"
  selfLink: /apis/extensions/v1beta1/namespaces/kubemark1/deployments/hollow-node
  uid: 1b11d8b2-ce03-11e8-95f8-525400c802e4
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      name: hollow-node
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        name: hollow-node
    spec:
      containers:
      - command:
        - /bin/sh
        - -c
        - /kubemark --morph=kubelet --name=$(NODE_NAME) --v=4 --kubeconfig=/etc/kubernetes/kubectl.kubeconfig
          --alsologtostderr 1>>/var/log/kubelet-$(NODE_NAME).log 2>&1
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: kubemark:v1
        imagePullPolicy: IfNotPresent
        name: hollow-kubelet
        ports:
        - containerPort: 4194
          protocol: TCP
        - containerPort: 10250
          protocol: TCP
        - containerPort: 10255
          protocol: TCP
        resources:
          requests:
            cpu: 60m
            memory: 100M
        securityContext:
          privileged: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/kubernetes
          name: kubeconfig-volume
          readOnly: true
        - mountPath: /var/log
          name: logs-volume
      - command:
        - /bin/sh
        - -c
        - /kubemark --morph=proxy --use-real-proxier=true --name=$(NODE_NAME) --v=4
          --kubeconfig=/etc/kubernetes/kubectl.kubeconfig --alsologtostderr 1>>/var/log/kubeproxy-$(NODE_NAME).log
          2>&1
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: kubemark:v1
        imagePullPolicy: IfNotPresent
        name: hollow-proxy
        resources:
          requests:
            cpu: 30m
            memory: 20M
        securityContext:
          privileged: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/kubernetes
          name: kubeconfig-volume
          readOnly: true
        - mountPath: /var/log
          name: logs-volume
      dnsPolicy: ClusterFirst
      initContainers:
      - command:
        - sysctl
        - -w
        - fs.inotify.max_user_instances=200
        image: busybox
        imagePullPolicy: IfNotPresent
        name: init-inotify-limit
        resources: {}
        securityContext:
          privileged: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      nodeSelector:
        node: py_node
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /var/log
          type: ""
        name: logs-volume
      - hostPath:
          path: /etc/kubernetes
          type: Directory
        name: kubeconfig-volume
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2018-10-12T09:42:16Z
    lastUpdateTime: 2018-10-12T09:42:16Z
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: 2018-10-12T09:42:16Z
    lastUpdateTime: 2018-10-12T13:13:05Z
    message: ReplicaSet "hollow-node-648bf9bb4" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 10
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

hollow-proxy container logs:

I1012 12:37:13.681139       7 proxier.go:638] syncProxyRules took 622.278µs
I1012 12:37:13.681156       7 asm_amd64.s:573] Shutting down service config controller
E1012 12:37:13.681249       7 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/root/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/root/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/root/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/root/go/src/k8s.io/kubernetes/vendor/k8s.io/utils/exec/testing/fake_exec.go:52
/root/go/src/k8s.io/kubernetes/pkg/util/conntrack/conntrack.go:61
/root/go/src/k8s.io/kubernetes/pkg/util/conntrack/conntrack.go:49
/root/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:1367
/root/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:532
/root/go/src/k8s.io/kubernetes/pkg/proxy/config/config.go:211
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2c68ad9]

goroutine 74 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/root/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x32dc9e0, 0x5c20d20)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/vendor/k8s.io/utils/exec/testing.(*FakeExec).LookPath(0xc420ad4450, 0x3ac6dba, 0x9, 0xc4208c51d0, 0x0, 0xc4200a4800, 0x4e7e83)
	/root/go/src/k8s.io/kubernetes/vendor/k8s.io/utils/exec/testing/fake_exec.go:52 +0x39
k8s.io/kubernetes/pkg/util/conntrack.Exec(0x3e3c000, 0xc420ad4450, 0xc4208c51d0, 0x5, 0x5, 0x5, 0x5)
	/root/go/src/k8s.io/kubernetes/pkg/util/conntrack/conntrack.go:61 +0x5e
k8s.io/kubernetes/pkg/util/conntrack.ClearEntriesForIP(0x3e3c000, 0xc420ad4450, 0xc420768c70, 0xc, 0x3abdd6d, 0x3, 0x0, 0x0)
	/root/go/src/k8s.io/kubernetes/pkg/util/conntrack/conntrack.go:49 +0x17d
k8s.io/kubernetes/pkg/proxy/iptables.(*Proxier).syncProxyRules(0xc420d8fe40)
	/root/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:1367 +0xb643
k8s.io/kubernetes/pkg/proxy/iptables.(*Proxier).OnServiceSynced(0xc420d8fe40)
	/root/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:532 +0x62
k8s.io/kubernetes/pkg/proxy/config.(*ServiceConfig).Run(0xc420ad5080, 0xc4200b62a0)
	/root/go/src/k8s.io/kubernetes/pkg/proxy/config/config.go:211 +0x183
created by k8s.io/kubernetes/cmd/kube-proxy/app.(*ProxyServer).Run
	/root/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:555 +0x283

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

boot up hollow-proxy container with "--use-real-proxier=true"

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

I found that in "kubemark/hollow-node.go" we init a 'FakeExec' struct, but we don't init LookPathFunc in 'FakeExec' struct.

https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/cmd/kubemark/hollow-node.go#L187

```
	if config.Morph == "proxy" {
		...
		sysctl := fakesysctl.NewFake()
		execer := &fakeexec.FakeExec{}
		eventBroadcaster := record.NewBroadcaster()
		...
		hollowProxy.Run()
	}
```

https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/utils/exec/testing/fake_exec.go#L28

```
// A simple scripted Interface type.
type FakeExec struct {
	CommandScript []FakeCommandAction
	CommandCalls  int
	LookPathFunc  func(string) (string, error)
}
```

https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/utils/exec/testing/fake_exec.go#L51

```
func (fake *FakeExec) LookPath(file string) (string, error) {
	return fake.LookPathFunc(file)
}
```

So,  execer.LookPathFunc is a vild pointer, so it cause panic.


[kubeproxy-hollow-node-85c74bd457-ctszb.log](https://github.com/kubernetes/kubernetes/files/2473437/kubeproxy-hollow-node-85c74bd457-ctszb.log)

--- a/cmd/kubemark/hollow-node.go
+++ b/cmd/kubemark/hollow-node.go
@@ -17,6 +17,7 @@ limitations under the License.
 package main
 
 import (
+	"errors"
 	goflag "flag"
 	"fmt"
 	"math/rand"
@@ -185,7 +186,9 @@ func run(config *hollowNodeConfig) {
 		}
 		iptInterface := fakeiptables.NewFake()
 		sysctl := fakesysctl.NewFake()
-		execer := &fakeexec.FakeExec{}
+		execer := &fakeexec.FakeExec{
+			LookPathFunc: func(_ string) (string, error) { return "", errors.New("fake execer") },
+		}
 		eventBroadcaster := record.NewBroadcaster()
 		recorder := eventBroadcaster.NewRecorder(legacyscheme.Scheme, v1.EventSource{Component: "kube-proxy", Host: config.NodeName})
 

